usersession/userd/settings.go

```diff
@@ -78,14 +95,102 @@ func allowedSetting(setting string) bool {
 }
 
 func settingWhitelisted(setting string) *dbus.Error {
+	// Trim off any subproperty from the check
+	settingTrimmed := strings.SplitN(string(setting), "/", 2)[0]
 	for _, whitelisted := range settingsWhitelist {
-		if setting == whitelisted {
+		if settingTrimmed == whitelisted {
 			return nil
 		}
 	}
 	return dbus.MakeFailedError(fmt.Errorf("cannot use setting %q: not allowed", setting))
 }
 
+func safeSnapFromSender(s *Settings, sender dbus.Sender) (string, *dbus.Error) {
+	// avoid information leak: see https://github.com/snapcore/snapd/pull/4073#discussion_r146682758
+	snap, err := snapFromSender(s.conn, sender)
+	if err != nil {
+		return "", dbus.MakeFailedError(err)
+	}
+	return snap, nil
+}
+
+func desktopFile(s *Settings, command string, setting string, desktopValue string, sender dbus.Sender) (string, *dbus.Error) {
+	snap, err := safeSnapFromSender(s, sender)
+	if err != nil {
+		return "", err
+	}
+	if err := settingWhitelisted(setting); err != nil {
+		return "", err
+	}
+
+	if !allowedSetting(desktopValue) {
+		return "", dbus.MakeFailedError(fmt.Errorf("cannot %q setting %q to value %q: value not allowed", command, setting, desktopValue))
+	}
+
+	// FIXME: this works only for desktop files
+	desktopFile := fmt.Sprintf("%s_%s", snap, desktopValue)
+	return desktopFile, nil
+}
+
+func desktopFileFromOutput(s *Settings, output string, sender dbus.Sender) (string, *dbus.Error) {
+	snap, err := safeSnapFromSender(s, sender)
+	if err != nil {
+		return "", err
+	}
+	if !strings.HasPrefix(output, snap+"_") {
+		return "NOT-THIS-SNAP.desktop", nil
+	}
+
+	desktopFile := strings.SplitN(output, "_", 2)[1]
+	return strings.TrimSpace(desktopFile), nil
+}
+
+func setDialog(s *Settings, setting string, desktopFile string, sender dbus.Sender) *dbus.Error {
+	df := filepath.Join(dirs.SnapDesktopFilesDir, desktopFile)
+	if !osutil.FileExists(df) {
+		return dbus.MakeFailedError(fmt.Errorf("cannot find desktop file %q", df))
+	}
+
+	// FIXME: we need to know the parent PID or our dialog may pop under
+	//        the existing windows. We might get it with the help of
+	//        the xdg-settings tool inside the core snap. It would have
+	//        to get the PID of the process asking for the settings
+	//        then xdg-settings can sent this to us and we can intospect
+	//        the X windows for _NET_WM_PID and use the windowID to
+	//        attach to zenity - not sure how this translate to the
+	//        wayland world though :/
+	dialog, uiErr := ui.New()
+	if uiErr != nil {
+		return dbus.MakeFailedError(fmt.Errorf("cannot ask for settings change: %v", uiErr))
+	}
+
+	snap, err := safeSnapFromSender(s, sender)
+	if err != nil {
+		return err
+	}
```

Comment:
This is redundant, no? Both Set() and SetSub() called desktopFile() and I don't see where `s` or `sender` was updated to need to recheck. Am I missing something?

Reply:
Perhaps, but it feels unsafe to leave an error return value unaddressed, in case this chunk of code is refactored around later.